src/plugins/api/index_pattern_endpoints.js

```diff
@@ -0,0 +1,221 @@
+        getMappings(pattern, client),
+        stitchPatternAndMappings
+      ).then(removeDeprecatedFieldProps)
+        .then(function (pattern) {
+          reply(pattern);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns',
+    method: 'POST',
+    handler: function (req, reply) {
+      const client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
+      const mappings = _.omit(_.mapValues(_.indexBy(req.payload.fields, 'name'), (value) => {
+        return value.mapping;
+      }), _.isUndefined);
+      indexPattern.fields = JSON.stringify(_.map(indexPattern.fields, (field) => {
+        return _.omit(field, 'mapping');
+      }));
+
+
+      client.create({
+        index: '.kibana',
+        type: 'index-pattern',
+        id: indexPattern.title,
+        body: indexPattern
+      }).then((patternResponse) => {
+        if (!isWildcard || _.isEmpty(mappings)) {
+          return patternResponse;
+        }
+        else {
+          return client.indices.exists({
+            index: indexPattern.title
+          }).then((matchingIndices) => {
+            if (matchingIndices) {
+              throw Boom.conflict('Cannot create an index template if existing indices already match index pattern');
+            }
+            else {
+              return client.indices.putTemplate({
+                order: 0,
+                create: true,
+                name: patternToTemplate(indexPattern.title),
+                body: {
+                  template: indexPattern.title,
+                  mappings: {
+                    _default_: {
+                      properties: mappings
+                    }
+                  }
+                }
+              });
+            }
+          }).catch((templateError) => {
+            return client.delete({
+              index: '.kibana',
+              type: 'index-pattern',
+              id: indexPattern.title
+            }).then(() => {
+              throw templateError;
+            }, () => {
+              throw new Error(`index-pattern ${indexPattern.title} created successfully but index template creation
+                failed. Failed to rollback index-pattern creation, must delete manually.`);
+            });
+          });
+        }
+      }).then(() => {
+        reply('success').statusCode = 201;
+      }).catch(function (error) {
+        reply(handleESError(error));
+      });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns/{id}',
+    method: 'PUT',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
```

Comment:
That was my initial reaction too, but the way I'm implementing it I think it is REST-y. The schema for the index pattern REST resource that I'm defining includes the (optional) mappings for the fields  (one item on my to-do list is to add a Joi schema and beef up the documentation around this in the ticket). Those mappings are stored in an index template, not in the .kibana index and that's ok. Different parts of a REST resource can be persisted in different places (e.g. in two separate tables in a SQL database). Creating the template isn't a side effect since the mappings are a part of the index pattern REST resource by definition. If a client POSTs an index pattern without mappings, no template will be created, so they have to intentionally opt-in to this behavior. 

That said, the idea to create an index template was Rashid's, so that's why I was trying to get his input :smile: